apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -87,50 +92,137 @@ public override int GetHashCode()
-                        {
-                            if (_client == null)
-                            {
-                                var channel = new Channel(
-                                    endpoint.Host,
-                                    endpoint.Port,
-                                    _key.Credential.ToChannelCredentials());
-                                _client = SpannerClient.Create(channel);
-                            }
-                        }
+                        _multiClientPool.RemoveAt(_multiClientPool.Count - 1);
                     }
-                    else
+                    while (_multiClientPool.Count < SpannerOptions.Instance.MaximumGrpcChannels)
                     {
-                        lock (_sync)
-                        {
-                            if (_creationTask == null || _creationTask.IsFaulted)
-                            {
-                                _creationTask = SpannerClient.CreateAsync(endpoint);
-                            }
-                        }
-                        // await needs to be done outside of the monitor.
-                        _client = await _creationTask.ConfigureAwait(false);
+                        var newEntry = new SpannerClientCreator(_key);
+                        _multiClientPool.Add(newEntry, newEntry);
                     }
+
+                    //now grab the first item in the sorted list, increment refcnt, re-sort and return.
+                    var first = _multiClientPool.First();
+                    result = first.Value.AcquireClientAsync();
+                    _multiClientPool.Remove(first.Key);
+                    _multiClientPool.Add(first.Key, first.Value);
+                }
+                return result;
+            }
+
+            public void ReleaseClient(SpannerClient client)
+            {
+                lock (_sync)
+                {
+                    //find the entry and release refcnt and re-sort
+                    var match = _multiClientPool.FirstOrDefault(x => x.Key.MatchesClient(client));
+                    if (match.Key != null)
+                    {
+                        match.Value.Release();
+                        _multiClientPool.Remove(match.Key);
+                        _multiClientPool.Add(match.Key, match.Value);
+                    }
+                }
+            }
+        }
+
+        private class SpannerClientCreator : IComparable<SpannerClientCreator>, IComparable
+        {
+            private  Lazy<Task<SpannerClient>> _creationTask;
+            private int _refCount = 0;
+            private ClientCredentialKey _key;
+
+            public SpannerClientCreator(ClientCredentialKey key)
+            {
+                _key = key;
+                _creationTask = new Lazy<Task<SpannerClient>>(AcquireClientImplAsync);
+            }
+
+            public bool MatchesClient(SpannerClient client) => _creationTask.IsValueCreated &&
```

Comment:
I've touched it up so that failed creates (which should only happen on bad credentials) resets the lazy to try again when the second call is made.
So while it will continue to throw if you continue to try to access a client, it will actually try multiple times -- which is probably better than blindly throwing the cached task exception from the first try.